packages/query-store/src/QueryStore.ts

```diff
@@ -0,0 +1,340 @@
+			}
+		}
+	};
+
+	/** Force the QueryStore to fetch data */
+	fetch = () => this.#fetchData();
+
+	#fetchData = () => {
+		if (this.#loading || this.#loaded) {
+			return;
+		}
+		this.#loading = true;
+		this.publish();
+
+		const queryWithComment = `--data\n${this.#query.toString()}`;
+
+		return handleMaybePromise<QueryResult[]>((result) => {
+			this.#values = result;
+			this.#loading = false;
+			this.#loaded = true;
+			return this.#fetchLength();
+		}, this.#exec(queryWithComment));
+	};
+
+	#fetchLength = () => {
+		if (this.#lengthLoading || this.#lengthLoaded) {
+			return;
+		}
+		this.#lengthLoading = true;
+		this.publish();
+
+		const countQuery = new Query()
+			.with({ original: this.#query })
+			.select({ length: count('*') })
+			.from('original');
+		const queryWithComment = `--len\n${countQuery}`;
+
+		return handleMaybePromise<QueryResult[]>((result) => {
+			const [row] = result;
+
+			this.#length = row.length as number;
+			this.#lengthLoaded = true;
+			this.#lengthLoading = false;
+
+			this.publish();
+		}, this.#exec(queryWithComment));
+	};
+
+	#fetchMetadata = () => {
+		return handleMaybePromise((queryResult: QueryResult[]) => {
+			this.#columns = queryResult.map((row) => ({
+				name: row.column_name!.toString(),
+				type: row.column_type!.toString()
+			}));
+			this.#mockResult = Object.fromEntries(this.#columns.map((c) => [c.name, null]));
+
+			this.publish();
+		}, this.#exec(`--col-metadata\nDESCRIBE ${this.#query.toString()}`));
+	};
+
+	/////////
+	// Builder Methods
+	/////////
+
+	/**
+	 * Shared cache of existing QueryStores to reduce the number of stores initialized
+	 */
+	private static cache: Record<string, QueryStore> = {};
+
+	/**
+	 * Array of child ids that the store is currently subscribed to.
+	 * @todo: need to clean up subscriptions if/when stores are gc'd
+	 */
+	#subscriptions: string[] = [];
+
+	/**
+	 * Wraps an derivation function with memoization provided by {@link QueryStore.cache}
+	 */
+	#withStoreCache = (aggKey: string, aggFunc: AggFunction, passCurrentAsInitial = false) => {
+		return (...args: unknown[]) => {
+			// Build a unique ID for the target child store
+			const newQuery = aggFunc(this.#query.clone(), ...args);
+
+			const id = buildId(newQuery.toString());
+
+			// If there is a root; subscription operations will function against that
+			// If there isn't; then this is a root
+			const subscriber = this.root ?? this;
+
+			// If caching is enabled and the id exists in the cache
+			if (id in QueryStore.cache && !this.opts.disableCache) {
+				// Use the cache
+				const cachedQuery = QueryStore.cache[id];
+
+				if (!subscriber.#subscriptions.includes(id)) {
+					cachedQuery.subscribe(subscriber.publish);
+					subscriber.#subscriptions.push(id);
+				}
+
+				return cachedQuery.#proxied;
+			}
+			// Construct a new store, subscribe and cache it
+			const newStore = new QueryStore(
+				newQuery,
+				this.#exec,
+				id,
+				{
+					...this.opts,
```

Comment:
For modifications like `where`, `orderby`, `limit` and `offset`; the columns don't change, so passing some initial data can help keep _something_ on the DOM / in the result set to minimize the amount of time the user experiences loading.

Reply:
So is it just for the column headers ? It seems like we are not actually applying these desired aggregation directly on the data set returned by `parent` at any point?

Reply:
> So is it just for the column headers ?

Right now; yes, this might be worth rethinking a bit in the future to instead pass column-metadata so `#mockResult` can be used